File: v4/rust/iprange-livedb/src/validation/tree.rs
use crate::contract::MAX_TREE_LEVEL;
use crate::error::Result;
use crate::mapping::{ByteRange, ByteSource, PageView};
use crate::slotted_page::LayoutInspection;

use super::context::Context;
use super::page::{self, TreePageSpec};
use super::{ValidationObject, ValidationReason, ValidationSink};

pub(crate) use crate::slotted_page::CellLayout;

pub(crate) trait Codec {
    type Key: Copy + Ord;

    const BRANCH_TYPE: u8;
    const LEAF_TYPE: u8;
    const AUX: u32;
    const BRANCH_LAYOUT: CellLayout;
    const LEAF_LAYOUT: CellLayout;
    const BRANCH_INVALID: ValidationReason = ValidationReason::TreeOrderInvalid;
    const LEAF_INVALID: ValidationReason;

    fn branch_key<P: ByteSource>(cell: P) -> Option<Self::Key>;
    fn branch_child<P: ByteSource>(cell: P) -> Option<u32>;
    fn leaf_key<P: ByteSource>(cell: P) -> Option<Self::Key>;
}

pub(crate) struct WalkResult {
    pub(crate) records: u64,
}

pub(crate) fn walk<'m, C, S, F>(
    context: &mut Context<'m, S>,
    root: u32,
    object: ValidationObject,
    mut leaf: F,
) -> Result<WalkResult>
where
    C: Codec,
    S: ValidationSink,
    F: FnMut(&mut Context<'m, S>, u32, ByteRange<PageView<'m>>) -> Result<()>,
{
    if root == 0 {
        return Ok(WalkResult { records: 0 });
    }
    let mut state = State::<C::Key> {
        records: 0,
        previous: None,
    };
    let mut path = [0; MAX_TREE_LEVEL as usize + 1];
    walk_node::<C, S, F>(
        context, root, object, None, true, &mut path, 0, &mut state, &mut leaf,
    )?;
    Ok(WalkResult {
        records: state.records,
    })
}

#[allow(clippy::too_many_arguments)]
fn walk_node<'m, C, S, F>(
    context: &mut Context<'m, S>,
    page_number: u32,
    object: ValidationObject,
    expected_level: Option<u16>,
    root: bool,
    path: &mut [u32; MAX_TREE_LEVEL as usize + 1],
    depth: usize,
    state: &mut State<C::Key>,
    leaf: &mut F,
) -> Result<Option<C::Key>>
where
    C: Codec,
    S: ValidationSink,
    F: FnMut(&mut Context<'m, S>, u32, ByteRange<PageView<'m>>) -> Result<()>,
{
    let Some(page) = read_node_page(context, page_number, object, path, depth)? else {
        state.previous = None;
        return Ok(None);
    };
    let Some(header) =
        read_node_header::<C, S>(context, page_number, page, object, expected_level)?
    else {
        state.previous = None;
        return Ok(None);
    };
    validate_root_shape(context, page_number, object, root, header)?;
    let layout = node_layout::<C>(header.level);
    let Some(cells) = validate_layout(context, page_number, page, object, header, layout)? else {
        state.previous = None;
        return Ok(None);
    };
    if header.level == 0 {
        walk_leaf::<C, S, F>(context, page_number, cells, object, state, leaf)
    } else {
        walk_branch::<C, S, F>(
            context,
            page_number,
            cells,
            header,
            object,
            path,
            depth,
            state,
            leaf,
        )
    }
}

fn read_node_page<'m, S: ValidationSink>(
    context: &mut Context<'m, S>,
    page_number: u32,
    object: ValidationObject,
    path: &mut [u32; MAX_TREE_LEVEL as usize + 1],
    depth: usize,
) -> Result<Option<PageView<'m>>> {
    let Some(slot) = path.get_mut(depth) else {
        context.emit(
            ValidationReason::TreeLevelInvalid,
            object,
            Some(page_number),
            None,
            None,
        )?;
        return Ok(None);
    };
    *slot = page_number;
    context.read_graph_page(page_number, object, &path[..depth])
}

fn read_node_header<C: Codec, S: ValidationSink>(
    context: &mut Context<'_, S>,
    page_number: u32,
    page: PageView<'_>,
    object: ValidationObject,
    expected_level: Option<u16>,
) -> Result<Option<page::SlottedHeader>> {
    page::slotted_header(
        context,
        page_number,
        page,
        object,
        TreePageSpec {
            branch_type: C::BRANCH_TYPE,
            leaf_type: C::LEAF_TYPE,
            aux: C::AUX,
            expected_level,
        },
    )
}

fn validate_root_shape<S: ValidationSink>(
    context: &mut Context<'_, S>,
    page_number: u32,
    object: ValidationObject,
    root: bool,
    header: page::SlottedHeader,
) -> Result<()> {
    if root && header.level > 0 && header.item_count == 1 {
        context.emit(
            ValidationReason::TreeLevelInvalid,
            object,
            Some(page_number),
            None,
            None,
        )?;
    }
    Ok(())
}

fn node_layout<C: Codec>(level: u16) -> CellLayout {
    if level == 0 {
        C::LEAF_LAYOUT
    } else {
        C::BRANCH_LAYOUT
    }
}

#[allow(clippy::too_many_arguments)]
fn walk_branch<'m, C, S, F>(
    context: &mut Context<'m, S>,
    page_number: u32,
    cells: LayoutInspection<PageView<'m>>,
    header: page::SlottedHeader,
    object: ValidationObject,
    path: &mut [u32; MAX_TREE_LEVEL as usize + 1],
    depth: usize,
    state: &mut State<C::Key>,
    leaf: &mut F,
) -> Result<Option<C::Key>>
where
    C: Codec,
    S: ValidationSink,
    F: FnMut(&mut Context<'m, S>, u32, ByteRange<PageView<'m>>) -> Result<()>,
{
    let mut first = None;
    let mut previous = None;
    for cell in cells.cells() {
        let Some((key, child)) =
            branch_entry::<C, S, _>(context, page_number, object, cell, state)?
        else {
            continue;
        };
        record_branch_key(context, page_number, object, key, &mut first, &mut previous)?;
        let actual = walk_node::<C, S, F>(
            context,
            child,
            object,
            Some(header.level - 1),
            false,
            path,
            depth + 1,
            state,
            leaf,
        )?;
        validate_fence(context, page_number, object, key, actual)?;
    }
    Ok(first)
}

fn branch_entry<C: Codec, S: ValidationSink, P: ByteSource>(
    context: &mut Context<'_, S>,
    page_number: u32,
    object: ValidationObject,
    cell: P,
    state: &mut State<C::Key>,
) -> Result<Option<(C::Key, u32)>> {
    let entry = C::branch_key(cell).zip(C::branch_child(cell));
    if entry.is_none() {
        context.emit(C::BRANCH_INVALID, object, Some(page_number), None, None)?;
        state.previous = None;
    }
    Ok(entry)
}

fn record_branch_key<K: Copy + Ord, S: ValidationSink>(
    context: &mut Context<'_, S>,
    page_number: u32,
    object: ValidationObject,
    key: K,
    first: &mut Option<K>,
    previous: &mut Option<K>,
) -> Result<()> {
    first.get_or_insert(key);
    if previous.is_some_and(|prior| prior >= key) {
        context.emit(
            ValidationReason::TreeOrderInvalid,
            object,
            Some(page_number),
            None,
            None,
        )?;
    }
    *previous = Some(key);
    Ok(())
}

fn validate_fence<K: Copy + PartialEq, S: ValidationSink>(
    context: &mut Context<'_, S>,
    page_number: u32,
    object: ValidationObject,
    expected: K,
    actual: Option<K>,
) -> Result<()> {
    if actual.is_some_and(|actual| actual != expected) {
        context.emit(
            ValidationReason::TreeFenceInvalid,
            object,
            Some(page_number),
            None,
            None,
        )?;
    }
    Ok(())
}

fn walk_leaf<'m, C, S, F>(
    context: &mut Context<'m, S>,
    page_number: u32,
    cells: LayoutInspection<PageView<'m>>,
    object: ValidationObject,
    state: &mut State<C::Key>,
    leaf: &mut F,
) -> Result<Option<C::Key>>
where
    C: Codec,
    S: ValidationSink,
    F: FnMut(&mut Context<'m, S>, u32, ByteRange<PageView<'m>>) -> Result<()>,
{
    let mut first = None;
    for cell in cells.cells() {
        walk_leaf_cell::<C, S, F>(context, page_number, cell, object, state, &mut first, leaf)?;
    }
    Ok(first)
}

fn walk_leaf_cell<'m, C, S, F>(
    context: &mut Context<'m, S>,
    page_number: u32,
    cell: ByteRange<PageView<'m>>,
    object: ValidationObject,
    state: &mut State<C::Key>,
    first: &mut Option<C::Key>,
    leaf: &mut F,
) -> Result<()>
where
    C: Codec,
    S: ValidationSink,
    F: FnMut(&mut Context<'m, S>, u32, ByteRange<PageView<'m>>) -> Result<()>,
{
    let Some(key) = C::leaf_key(cell) else {
        context.emit(C::LEAF_INVALID, object, Some(page_number), None, None)?;
        state.previous = None;
        return Ok(());
    };
    first.get_or_insert(key);
    if state.previous.is_some_and(|previous| previous >= key) {
        context.emit(
            ValidationReason::TreeOrderInvalid,
            object,
            Some(page_number),
            None,
            None,
        )?;
    }
    state.previous = Some(key);
    state.records = state
        .records
        .checked_add(1)
        .ok_or(crate::error::Error::ArithmeticOverflow(
            "validation tree record count",
        ))?;
    leaf(context, page_number, cell)
}

struct State<K> {
    records: u64,
    previous: Option<K>,
}

fn validate_layout<S: ValidationSink, P: ByteSource>(
    context: &mut Context<'_, S>,
    page_number: u32,
    page: P,
    object: ValidationObject,
    header: page::SlottedHeader,
    layout: CellLayout,
) -> Result<Option<LayoutInspection<P>>> {
    match layout {
        CellLayout::Fixed(length) => {
            page::validate_fixed_cells(context, page_number, page, object, header, length)
        }
        CellLayout::Variable { minimum, maximum } => page::validate_variable_cells(
            context,
            page_number,
            page,
            object,
            header,
            minimum,
            maximum,
        ),
    }
}

File: v4/go/internal/validation/tree.go
package validation

// Tree walk validation (Rust validation/tree.rs): one codec-driven walk
// over a committed tree root with root-shape, per-page order, and fence
// checks, streaming findings through the context. The walk is
// allocation-free: the first-key results and the ordered-key cursor are
// values, and the cell iterator slices the inspected page.

import (
	"github.com/firehol/iprange/v4/go/internal/format"
	"github.com/firehol/iprange/v4/go/internal/tree"
)

// treeCodec carries the per-tree wire decoding of one validated B+tree
// (Rust validation/tree.rs Codec). The decode functions return ok=false
// for an undecodable cell; the walk reports the codec's invalid class.
type treeCodec struct {
	branchType    byte
	leafType      byte
	aux           uint32
	branchLayout  format.CellLayout
	leafLayout    format.CellLayout
	branchInvalid ValidationReason
	leafInvalid   ValidationReason
	branchKey     func(cell []byte) (tree.Key, bool)
	branchChild   func(cell []byte) (uint32, bool)
	leafKey       func(cell []byte) (tree.Key, bool)
}

// treeWalkResult mirrors Rust WalkResult.
type treeWalkResult struct {
	records uint64
}

// treeWalkState mirrors the Rust State: the ordered previous key and the
// record count of the whole walk.
type treeWalkState struct {
	records     uint64
	previous    tree.Key
	hasPrevious bool
}

// walkTree mirrors Rust tree::walk over one root: a zero root is an empty
// tree, every leaf record runs leaf, and the result carries the decoded
// record count.
func walkTree(ctx *context, root uint32, object ValidationObject, codec treeCodec, leaf func(*context, uint32, []byte) error) (treeWalkResult, error) {
	if root == 0 {
		return treeWalkResult{}, nil
	}
	state := treeWalkState{}
	var path [format.MaxTreeLevel + 1]uint32
	if _, _, err := walkTreeNode(ctx, root, object, codec, nil, true, &path, 0, &state, leaf); err != nil {
		return treeWalkResult{}, err
	}
	return treeWalkResult{records: state.records}, nil
}

// walkTreeNode visits one tree node and returns the first key of its
// subtree for the parent fence check (Rust walk_node).
func walkTreeNode(ctx *context, pageNumber uint32, object ValidationObject, codec treeCodec, expectedLevel *uint16, root bool, path *[format.MaxTreeLevel + 1]uint32, depth int, state *treeWalkState, leaf func(*context, uint32, []byte) error) (tree.Key, bool, error) {
	page, err := readTreeNodePage(ctx, pageNumber, object, path, depth)
	if err != nil || page == nil {
		state.hasPrevious = false
		return tree.Key{}, false, err
	}
	header, err := treePageHeader(ctx, pageNumber, page, object, treePageSpec{
		branchType:    codec.branchType,
		leafType:      codec.leafType,
		aux:           codec.aux,
		expectedLevel: expectedLevel,
	})
	if err != nil || header == nil {
		state.hasPrevious = false
		return tree.Key{}, false, err
	}
	if err := validateRootShape(ctx, pageNumber, object, root, header); err != nil {
		return tree.Key{}, false, err
	}
	inspection, err := validateTreeLayout(ctx, pageNumber, page, object, header, nodeLayout(codec, header.Level))
	if err != nil || inspection == nil {
		state.hasPrevious = false
		return tree.Key{}, false, err
	}
	if header.Level == 0 {
		return walkTreeLeaf(ctx, pageNumber, inspection, object, codec, state, leaf)
	}
	return walkTreeBranch(ctx, pageNumber, inspection, header, object, codec, path, depth, state, leaf)
}

// readTreeNodePage records the page in the walk path and reads it through
// the graph claims (Rust read_node_page: a path deeper than the maximum
// tree level is the TreeLevelInvalid class).
func readTreeNodePage(ctx *context, pageNumber uint32, object ValidationObject, path *[format.MaxTreeLevel + 1]uint32, depth int) ([]byte, error) {
	if depth >= len(path) {
		if err := ctx.emit(ReasonTreeLevelInvalid, object, &pageNumber, nil, nil); err != nil {
			return nil, err
		}
		return nil, nil
	}
	path[depth] = pageNumber
	return ctx.readGraphPage(pageNumber, object, path[:depth])
}

// validateRootShape mirrors tree.rs validate_root_shape: a root branch
// with exactly one record is the TreeLevelInvalid class.
func validateRootShape(ctx *context, pageNumber uint32, object ValidationObject, root bool, header *format.PageHeader) error {
	if root && header.Level > 0 && header.ItemCount == 1 {
		if err := ctx.emit(ReasonTreeLevelInvalid, object, &pageNumber, nil, nil); err != nil {
			return err
		}
	}
	return nil
}

// nodeLayout selects the cell layout of one tree level (Rust node_layout:
// level zero uses the leaf layout, every higher level the branch layout).
func nodeLayout(codec treeCodec, level uint16) format.CellLayout {
	if level == 0 {
		return codec.leafLayout
	}
	return codec.branchLayout
}

// walkTreeBranch visits every branch cell in order: the per-page key
// order, the child subtree with its expected level, and the child-first
// fence (Rust walk_branch).
func walkTreeBranch(ctx *context, pageNumber uint32, inspection *format.LayoutInspection, header *format.PageHeader, object ValidationObject, codec treeCodec, path *[format.MaxTreeLevel + 1]uint32, depth int, state *treeWalkState, leaf func(*context, uint32, []byte) error) (tree.Key, bool, error) {
	var keys branchKeys
	expected := header.Level - 1
	cells := inspection.Cells()
	for {
		cell, ok := cells.Next()
		if !ok {
			break
		}
		key, child, ok, err := branchTreeEntry(ctx, pageNumber, object, codec, cell, state)
		if err != nil || !ok {
			if err != nil {
				return tree.Key{}, false, err
			}
			continue
		}
		if err := recordBranchKey(ctx, pageNumber, object, key, &keys); err != nil {
			return tree.Key{}, false, err
		}
		actual, hasActual, err := walkTreeNode(ctx, child, object, codec, &expected, false, path, depth+1, state, leaf)
		if err != nil {
			return tree.Key{}, false, err
		}
		if err := validateFence(ctx, pageNumber, object, key, actual, hasActual); err != nil {
			return tree.Key{}, false, err
		}
	}
	return keys.first, keys.hasFirst, nil
}

// branchKeys is the ordered branch-key cursor of one branch page (Rust
// first/previous locals).
type branchKeys struct {
	first       tree.Key
	hasFirst    bool
	previous    tree.Key
	hasPrevious bool
}

// branchTreeEntry decodes one branch cell (Rust branch_entry: an
// undecodable key or child reports the codec branch-invalid class and
// resets the walk order state).
func branchTreeEntry(ctx *context, pageNumber uint32, object ValidationObject, codec treeCodec, cell []byte, state *treeWalkState) (tree.Key, uint32, bool, error) {
	key, keyOK := codec.branchKey(cell)
	child, childOK := codec.branchChild(cell)
	if !keyOK || !childOK {
		if err := ctx.emit(codec.branchInvalid, object, &pageNumber, nil, nil); err != nil {
			return tree.Key{}, 0, false, err
		}
		state.hasPrevious = false
		return tree.Key{}, 0, false, nil
	}
	return key, child, true, nil
}

// recordBranchKey records one branch key and reports a non-increasing
// key as the TreeOrderInvalid class (Rust record_branch_key).
func recordBranchKey(ctx *context, pageNumber uint32, object ValidationObject, key tree.Key, keys *branchKeys) error {
	if !keys.hasFirst {
		keys.first = key
		keys.hasFirst = true
	}
	if keys.hasPrevious && !keys.previous.Less(key) {
		if err := ctx.emit(ReasonTreeOrderInvalid, object, &pageNumber, nil, nil); err != nil {
			return err
		}
	}
	keys.previous = key
	keys.hasPrevious = true
	return nil
}

// validateFence checks one branch key against the first key of its child
// subtree (Rust validate_fence: a mismatch is the TreeFenceInvalid
// class).
func validateFence(ctx *context, pageNumber uint32, object ValidationObject, expected tree.Key, actual tree.Key, hasActual bool) error {
	if hasActual && !actual.Equal(expected) {
		if err := ctx.emit(ReasonTreeFenceInvalid, object, &pageNumber, nil, nil); err != nil {
			return err
		}
	}
	return nil
}

// walkTreeLeaf visits every leaf cell in order (Rust walk_leaf: the
// per-page key order, the checked record count, and the leaf callback).
func walkTreeLeaf(ctx *context, pageNumber uint32, inspection *format.LayoutInspection, object ValidationObject, codec treeCodec, state *treeWalkState, leaf func(*context, uint32, []byte) error) (tree.Key, bool, error) {
	first := tree.Key{}
	hasFirst := false
	cells := inspection.Cells()
	for {
		cell, ok := cells.Next()
		if !ok {
			break
		}
		key, ok := codec.leafKey(cell)
		if !ok {
			if err := ctx.emit(codec.leafInvalid, object, &pageNumber, nil, nil); err != nil {
				return tree.Key{}, false, err
			}
			state.hasPrevious = false
			continue
		}
		if !hasFirst {
			first = key
			hasFirst = true
		}
		if state.hasPrevious && !state.previous.Less(key) {
			if err := ctx.emit(ReasonTreeOrderInvalid, object, &pageNumber, nil, nil); err != nil {
				return tree.Key{}, false, err
			}
		}
		state.previous = key
		state.hasPrevious = true
		if state.records == ^uint64(0) {
			return tree.Key{}, false, &format.Error{Code: format.CodeArithmeticOverflow, Detail: "validation tree record count"}
		}
		state.records++
		if err := leaf(ctx, pageNumber, cell); err != nil {
			return tree.Key{}, false, err
		}
	}
	return first, hasFirst, nil
}
